geomstats/geometry/riemannian_metric.py

```diff
@@ -484,3 +438,61 @@ def sectional_curvature(
         normalization_factor = gs.where(
             condition, EPSILON, normalization_factor)
         return gs.where(~condition, sectional / normalization_factor, 0.)
+
+
+def loss(y_pred, y_true, metric):
+    """Compute loss function between prediction and ground truth.
+
+    Loss function given by a Riemannian metric,
+    expressed as the squared geodesic distance between the prediction
+    and the ground truth.
+
+    Parameters
+    ----------
+    y_pred : array-like, shape=[..., dim]
+        Prediction.
+    y_true : array-like, shape=[..., dim]
+        Ground-truth.
+    metric : RiemannianMetric
+        Metric.
+
+    Returns
+    -------
+    sq_dist : array-like, shape=[...,]
+        Loss, i.e. the squared distance.
+    """
+    sq_dist = metric.squared_dist(y_pred, y_true)
+    return sq_dist
+
+
+def grad(y_pred, y_true, metric):
+    """Closed-form for the gradient of the loss function.
+
+    Parameters
+    ----------
+    y_pred : array-like, shape=[..., dim]
+        Prediction.
+    y_true : array-like, shape=[..., dim]
+        Ground-truth.
+    metric : RiemannianMetric
+        Metric.
+
+    Returns
+    -------
+    loss_grad : array-like, shape=[...,]
+        Gradient of the loss.
+
+    """
+    tangent_vec = metric.log(base_point=y_pred, point=y_true)
+    grad_vec = - 2. * tangent_vec
+
+    inner_prod_mat = metric.metric_matrix(base_point=y_pred)
+    is_vectorized = inner_prod_mat.ndim == 3
+    axes = (0, 2, 1) if is_vectorized else (1, 0)
+
+    loss_grad = gs.einsum(
+        '...i,...ij->...i',
+        grad_vec,
+        gs.transpose(inner_prod_mat, axes=axes))
+
+    return loss_grad
```

Comment:
First, I never understood why these methods were here, what is their purpose?
Second, it would be much more useful to have one single method that returns both loss and grad, to avoid computing the loss twice!
Finally, why do we need the contraction with the metric matrix?

Reply:
It's some legacy code coming from a deep learning pipeline predicting on SO(3) - haha. Yes, they should go into a notebook to showcase this usecase, or be deleted.